src/query/functions/aggregation/base.go

```diff
@@ -0,0 +1,203 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package aggregation
+
+import (
+	"fmt"
+
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/executor/transform"
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/parser"
+)
+
+// NodeParams contains additional parameters required for aggregation ops
+type NodeParams struct {
+	Matching []string
+	Without  bool
+}
+
+type withKeysID func(tags models.Tags, matching []string) uint64
+
+func includeKeysID(tags models.Tags, matching []string) uint64 {
+	return tags.IDWithKeys(matching...)
+}
+
+func excludeKeysID(tags models.Tags, matching []string) uint64 {
+	return tags.IDWithExcludes(matching...)
+}
+
+type withKeysTags func(tags models.Tags, matching []string) models.Tags
+
+func includeKeysTags(tags models.Tags, matching []string) models.Tags {
+	return tags.TagsWithKeys(matching)
+}
+
+func excludeKeysTags(tags models.Tags, matching []string) models.Tags {
+	return tags.TagsWithoutKeys(matching)
+}
+
+// create the output, by tags,
+// returns a list of seriesMeta for the combined series,
+// and a list of [index lists].
+// Input series that exist in an index list are mapped to the
+// relevant index in the combined series meta.
+func collectSeries(params NodeParams, opName string, metas []block.SeriesMeta) ([][]int, []block.SeriesMeta) {
+	without, matching := params.Without, params.Matching
+
+	var idFunc withKeysID
+	var tagsFunc withKeysTags
+	if without {
+		idFunc = excludeKeysID
+		tagsFunc = excludeKeysTags
+	} else {
+		idFunc = includeKeysID
+		tagsFunc = includeKeysTags
+	}
+
+	type tagMatch struct {
+		indices []int
+		tags    models.Tags
+	}
+
+	tagMap := make(map[uint64]*tagMatch)
+	for i, meta := range metas {
+		id := idFunc(meta.Tags, matching)
+		if val, ok := tagMap[id]; ok {
+			val.indices = append(val.indices, i)
+		} else {
+			tagMap[id] = &tagMatch{
+				indices: []int{i},
+				tags:    tagsFunc(meta.Tags, matching),
+			}
+		}
+	}
+
+	collectedIndices := make([][]int, len(tagMap))
```

Comment:
Thoughts on renaming to `groupedIndices`, etc. instead?

Reply:
Makes more sense